ABAC (#2728)
neo4j/docs-operations: modules/ROOT/pages/authentication-authorization/attribute-based-access-control.adoc

:description: How to use Cypher to manage attribute-based access control on a graph.

////
[source, cypher, role=test-setup]
----
CREATE ROLE salesTeam;
CREATE ROLE engineeringTeamUK;
CREATE ROLE countryAccessRole;
----
////

:page-role: enterprise-edition aura-db-business-critical aura-db-dedicated

[[attribute-based-access-control]]
= Attribute-based access control

> Attribute-based access control, ABAC grants roles based on the evaluation of attributes (or claims) contained in a user's authentication token.

Reviewer: ```suggestion
Attribute-based access control (ABAC) grants roles based on the evaluation of attributes (or claims) contained in a user's authentication token.
```

We could subsequently use "ABAC" throughout?

Author: hmm, do we want to? Yes, if we want people to call it ABAC. No if we want to be clear and use the full name (I hate to stop reading and look up what an acronym stands for).

Author: Picked Phils suggestion

== Setup

Use the following steps to set up attribute-based access control:

1. Enable attribute-based access control in the `neo4j.conf` file by setting the `internal.dbms.feature_flag.attribute_based_access_control` setting to `true`.

> 2. Specify which OIDC provider will be used for attribute-based access control by setting the `internal.dbms.security.abac_enabled_authorization_providers` setting to the appropriate providers.

Reviewer: ```suggestion
2. Specify which OIDC provider(s) will be used for attribute-based access control by setting the `internal.dbms.security.abac_enabled_authorization_providers` setting to the appropriate providers.
```

The SET ENABLED clause is optional and can be used to enable or disable the rule upon creation. By default, the rule is enabled.

[NOTE]
====
To be able to create auth rules, the user must have the `AUTH RULE MANAGEMENT` privilege.
====

== Drop auth rules
To drop an existing authorization rule, use the following syntax:

[source, cypher25, role="noheader"]
----
DROP AUTH RULE ruleName [IF EXISTS];
----

[NOTE]
====
To be able to drop auth rules, the user must have the `AUTH RULE MANAGEMENT` privilege.
====

== Grant roles to auth rules
To specify which roles will be granted to the user when the authorization rule is fulfilled, use the `GRANT ROLE` command with the following syntax:

[source, cypher25, role="noheader"]
----
GRANT ROLE[S] role[, ...] TO AUTH RULE[S] ruleName[, ...];
----

For information about the `GRANT ROLE` command, see xref:authentication-authorization/manage-roles.adoc#access-control-assign-roles[Assigning roles to users].

[NOTE]
====
No roles with associated deny privileges can be assigned using attribute-based access control.
====

[NOTE]
====
To be able to grant roles, the user must have `GRANT ROLE` privilege.
====

> 

Reviewer: ```suggestion

[NOTE]
====
OR semantics are used between roles granted via ABAC and those granted by RBAC mechanisms. For example, given an ABAC rule which grants the `admin` role between the hours of 1900 and 2000, then a user who receives the `admin` role because it appears in their JWT `groups` claim (RBAC) will always have the `admin` role, even outside the hours of 1900 and 2000.  
====

```

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE salesRule
    SET CONDITION abac.oidc.user_attribute('department') = 'sales';
GRANT ROLE salesTeam TO AUTH RULE salesRule;
----

Granting the role `engineeringTeamUK` to users with the attribute `department` equal to `engineering` and `location` equal to `UK` can be done as follows:

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE engineeringUKRule
    SET CONDITION abac.oidc.user_attribute('department') = 'engineering'
        AND abac.oidc.user_attribute('location') = 'UK';
GRANT ROLE engineeringTeamUK TO AUTH RULE engineeringUKRule;
----

Granting a role based on the presence of the allowed countries in a list of citizenship countries in the claims can be done as follows:

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE ruleset_countries SET CONDITION
  any(country IN abac.oidc.user_attribute('citizenshipCountries')
    WHERE country IN ['c1', 'c5']);
GRANT ROLE countryAccessRole TO AUTH RULE ruleset_countries;
----

== Caveats and limitations

* When evaluating `abac.oidc.user_attribute('<claim_key>')`, if the claim does not exist in the authentication token, it will evaluate to `NULL`.

> * Newly created auth rules are not applied to existing user sessions. Users must re-authenticate to have the new rules applied.

Reviewer: ```suggestion
* Newly created auth rules are applied to existing user sessions, but will only have access to the user claims which had rules containing them at the start of the session. So only user claims which are used in pre-existing auth rules at the start of a session are retained. Users must re-authenticate to have new rules applied if the new rules use new claims.
```

[[attribute-based-access-control]]
= Attribute-based access control

Attribute-based access control, ABAC grants roles based on the evaluation of attributes (or claims) contained in a user's authentication token.

== Setup

Use the following steps to set up attribute-based access control:

1. Enable attribute-based access control in the `neo4j.conf` file by setting the `internal.dbms.feature_flag.attribute_based_access_control` setting to `true`.
2. Specify which OIDC provider(s) will be used for attribute-based access control by setting the `internal.dbms.security.abac_enabled_authorization_providers` setting to the appropriate providers.
3. Define the authorization rules that will be used to grant roles based on user attributes by using the Cypher command `CREATE AUTH RULE`, explained below.
4. Specify which roles will be assigned when the authorization rules are fulfilled using the `GRANT ROLE` command, explained below.

== Create auth rules
To create an authorization rule, use the following syntax:

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE ruleName [IF NOT EXISTS]
    SET CONDITION conditionExpression
    [ SET ENABLED {true | false} ]
----

[NOTE]
====

> To be able to create auth rules, the user must have the `AUTH RULE MANAGEMENT` privilege.

Reviewer 2: We should document this privilege and link to it from all the notes that mentions it

====

The SET ENABLED clause is optional and can be used to enable or disable the rule upon creation. By default, the rule is enabled.

If specifying `IF NOT EXISTS`, no error is raised if a rule with the same name already exists and the command has no effect.

The conditionExpression is a boolean cypher expression that evaluates user attributes (claims) contained in the authentication token.
To access an attribute, use the syntax `abac.oidc.user_attribute('<claim_key>')`.

> The conditionExpression can use any valid Cypher expression that evaluates to a boolean value, including logical operators (AND, OR, NOT etc), comparison operators (=, <>, <, >, <=, >=) and functions (e.g. `IN`, `ANY`, `ALL`, etc). The expression can also use the following in-build Neo4j functions:

Reviewer 2: The first sentence is long enough, no need to add to the line length by also having a second sentence there XD
```suggestion
The conditionExpression can use any valid Cypher expression that evaluates to a boolean value, including logical operators (AND, OR, NOT etc), comparison operators (=, <>, <, >, <=, >=) and functions (e.g. `IN`, `ANY`, `ALL`, etc).
The expression can also use the following in-build Neo4j functions:
```
(we tend to break separate sentences unto separate lines for easier reading/maintaining of the code even if it doesn't make a difference in the rendered docs)

> The conditionExpression can use any valid Cypher expression that evaluates to a boolean value, including logical operators (AND, OR, NOT etc), comparison operators (=, <>, <, >, <=, >=) and functions (e.g. `IN`, `ANY`, `ALL`, etc). The expression can also use the following in-build Neo4j functions:

Author: Maybe, what do you think @Hunterness and @phil198 ?

```
The conditionExpression can use any valid Cypher expression that evaluates to a boolean value, including logical operators (AND, OR, NOT etc), comparison operators (=, <>, <, >, <=, >=) and the following in-build Neo4j functions:

|===

[NOTE]
====
The functions, `date`, `datetime`, `localdatetime`, `localtime` and `time` are only supported when an input argument is used.
====

== Drop auth rules
To drop an existing authorization rule, use the following syntax:

[source, cypher25, role="noheader"]
----
DROP AUTH RULE ruleName [IF EXISTS]
----

If specifying `IF EXISTS`, no error is raised if the rule does not exist and the command has no effect.

[NOTE]
====
To be able to drop auth rules, the user must have the `AUTH RULE MANAGEMENT` privilege.
====

[[assign-roles-auth-rules]]
== Assign roles to auth rules
To specify which roles will be granted to the user when the authorization rule is fulfilled, use the `GRANT ROLE` command with the following syntax:

[source, cypher25, role="noheader"]
----
GRANT ROLE[S] role[, ...] TO AUTH RULE[S] ruleName[, ...]
----

> For information about the `GRANT ROLE` command, see xref:authentication-authorization/manage-roles.adoc#access-control-assign-roles[Assigning roles to users].

Reviewer 2: Same for revoke below

====

[NOTE]
====
OR semantics are used between roles granted via ABAC and those granted by RBAC mechanisms. For example, given an ABAC rule which grants the `admin` role between the hours of 1900 and 2000, then a user who receives the `admin` role because it appears in their JWT `groups` claim (RBAC) will always have the `admin` role, even outside the hours of 1900 and 2000.  
====

[[revoke-roles-auth-rules]]
== Revoke roles from auth rules
To revoke roles from an authorization rule, use the `REVOKE ROLE` command with the following syntax:

[source, cypher25, role="noheader"]
----
REVOKE ROLE[S] role[, ...] FROM AUTH RULE[S] ruleName[, ...]
----

For information about the `REVOKE ROLE` command, see xref:authentication-authorization/manage-roles.adoc#access-control-revoke-roles[Revoking roles from users].

[NOTE]
====
To be able to revoke roles, the user must have the xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-role-removal[`REMOVE ROLE`] privilege.
====

> == Examples

Reviewer 2: - [ ] Do we want any examples of DROP or REVOKE?
- [ ] Do we allow parameters anywhere? If so, should we have some examples of that? (both for auth rule names but also inside the conditions to show when they would be resolved (at grant time not evaluation time I'd assume)

    SET CONDITION abac.oidc.user_attribute('department') = 'engineering'
        AND abac.oidc.user_attribute('location') = 'UK';
GRANT ROLE engineeringTeamUK TO AUTH RULE engineeringUKRule;
----

Granting a role based on the presence of the allowed countries in a list of citizenship countries in the claims can be done as follows:

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE ruleset_countries SET CONDITION
  any(country IN abac.oidc.user_attribute('citizenshipCountries')
    WHERE country IN ['c1', 'c5']);
GRANT ROLE countryAccessRole TO AUTH RULE ruleset_countries;
----

Temporarily granting a role based on the value of a claim and the time being between 1900 and 2000 can be done as follows:

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE temporary_reader
    SET CONDITION abac.oidc.user_attribute('jobTitle') = 'cover'
        AND datetime.transaction().hour >= 19 AND datetime.transaction().hour < 20;
GRANT ROLE reader TO AUTH RULE temporary_reader;
----

== Caveats and limitations

* When evaluating `abac.oidc.user_attribute('<claim_key>')`, if the claim does not exist in the authentication token, it will evaluate to `NULL`.

> * Newly created auth rules are applied to existing user sessions, but will only have access to the user claims which had rules containing them at the start of the session. Only user claims which are used in pre-existing auth rules at the start of a session are retained. Users must re-authenticate to have new rules applied if the new rules use new claims.

Reviewer 2: Same as my other comment above about splitting long lines on sentences
```suggestion
* Newly created auth rules are applied to existing user sessions, but will only have access to the user claims which had rules containing them at the start of the session.
Only user claims which are used in pre-existing auth rules at the start of a session are retained.
Users must re-authenticate to have new rules applied if the new rules use new claims.
```

Might need `+` to keep things as a single paragraph with correct indenting when in a list 🤔 Don't remember but I think we can find other places which have similar things (might only be if you have multiple paragraphs so new lines in between 🤔 or even maybe on on said new lines 🤔 I don't remember but if the formatting isn't correct with the above version see if you can find what other lists do :P)
```suggestion
* Newly created auth rules are applied to existing user sessions, but will only have access to the user claims which had rules containing them at the start of the session.
+Only user claims which are used in pre-existing auth rules at the start of a session are retained.
+Users must re-authenticate to have new rules applied if the new rules use new claims.
```

[source, cypher25, role="noheader"]
----
CREATE AUTH RULE ruleName [IF NOT EXISTS]
    SET CONDITION conditionExpression
    [ SET ENABLED {true | false} ]
----

[NOTE]
====
To be able to create auth rules, the user must have the `AUTH RULE MANAGEMENT` privilege,
see xref:authentication-authorization/dbms-administration/dbms-auth-rule-management-privileges.adoc[DBMS AUTH RULE MANAGEMENT privileges].
====

The SET ENABLED clause is optional and can be used to enable or disable the rule upon creation. By default, the rule is enabled.

If specifying `IF NOT EXISTS`, no error is raised if a rule with the same name already exists and the command has no effect.

The conditionExpression is a boolean cypher expression that evaluates user attributes (claims) contained in the authentication token.
To access an attribute, use the syntax `abac.oidc.user_attribute('<claim_key>')`.

The conditionExpression can use any valid Cypher expression that evaluates to a boolean value, including logical operators (AND, OR, NOT etc), comparison operators (=, <>, <, >, <=, >=) and the following in-build Neo4j functions:

[cols="1,3",options="header"]
|===

> | Function group | Functions

Reviewer 2: is the ```functions (e.g. `IN`, `ANY`, `ALL`, etc)``` included in this list or do they need to be added? 🤔

Author: all() and any() is included. `IN` is not a function. it is an List operator https://neo4j.com/docs/cypher-manual/current/expressions/predicates/list-operators/
